# DS-SF-38 | 03 | `pandas` | Assignment | Starter Code

## `pandas` and Bistro

You've just told one of your friend that you are taking a Data Science class.  (Yeah!)  Your friend is running a bistro, a small restaurant, serving moderately priced simple meals in a modest setting ([Wikipedia](https://en.wikipedia.org/wiki/Bistro)).  She collected over some period of time the following information of her patrons' visits.

| Variable's name | Its meaning |
|:---:|:---|
| `name` | Patron's first name |
| `gender` | Patron's gender |
| `is_smoker` | Whether the patron is smoking or not |
| `party` | Party's size |
| `check` | Check amount (\$) (after taxes but before tips) |
| `tip` | Tip (\$) that the patron added to the check |
| `day` | Week day of the visit |
| `time` | Rough time estimate of the visit |

In this assignment, we will be exploring this dataset using `pandas`.<sup>(*)</sup>

<sup>(*)</sup> this dataset was adapted from the `tips` dataset of the `seaborn` package (https://github.com/mwaskom/seaborn-data)

> ### Question 1.  Import `pandas` (as `pd`).

In [51]:
import os

import numpy as np

# TODO
import pandas as pd
pd.set_option('display.max_rows', 10)
pd.set_option('display.notebook_repr_html', True)
pd.set_option('display.max_columns', 10)

> ### Question 2.  Read the `dataset-03-bistro.csv` dataset.

In [52]:
# TODO
df = pd.read_csv(os.path.join('../datasets/dataset-03-bistro.csv'))
df

day    time      name  gender  is_smoker  party  check   tip
0      Sunday  Dinner  Kimberly  Female      False      2  16.99  1.01
1      Sunday  Dinner  Nicholas    Male      False      3  10.34  1.66
2      Sunday  Dinner     Larry    Male      False      3  21.01  3.50
3      Sunday  Dinner    Joseph    Male      False      2  23.68  3.31
4      Sunday  Dinner    Janice  Female      False      4  24.59  3.61
..        ...     ...       ...     ...        ...    ...    ...   ...
239  Saturday  Dinner     Kevin    Male      False      3  29.03  5.92
240  Saturday  Dinner    Sandra  Female      False      2  27.18  2.00
241  Saturday  Dinner      Carl    Male      False      2  22.67  2.00
242  Saturday  Dinner       Jon    Male      False      2  17.82  1.75
243  Thursday  Dinner    Brandi  Female      False      2  18.78  3.00

[244 rows x 8 columns]

> ### Question 3.  What is the class of the `pandas` object storing the dataset?

In [53]:
# TODO
type(df)

pandas.core.frame.DataFrame

Answer: Stored as dataframe

> ### Question 4.  How many samples (i.e., rows) are in this dataset?

In [54]:
# TODO
df.shape[0]

244

Answer: 244

> ### Question 5.  How many variables (i.e., columns) are in this dataset?

In [55]:
# TODO
df.shape[1]

8

Answer: TODO

> ### Question 6.  Print the name of each column in the dataset, one name per line.

In [56]:
# TODO

# df.dtypes

for column in df.columns:
    print column


day
time
name
gender
is_smoker
party
check
tip


> ### Question 7.  Print the first two rows of the dataset to the console.  What does the output look like?

In [29]:
# TODO
df[:2]

day    time      name  gender  is_smoker  party  check   tip
0  Sunday  Dinner  Kimberly  Female      False      2  16.99  1.01
1  Sunday  Dinner  Nicholas    Male      False      3  10.34  1.66

Answer: TODO

> ### Question 8.  Extract the last 2 rows of the data frame and print them to the console.  What does the output look like?

In [30]:
# TODO
df[-2:]

day    time    name  gender  is_smoker  party  check   tip
242  Saturday  Dinner     Jon    Male      False      2  17.82  1.75
243  Thursday  Dinner  Brandi  Female      False      2  18.78  3.00

Answer: TODO

> ### Question 9.  Does the dataset contain any missing values?

In [31]:
# TODO
df.isnull().sum()

day          0
time         0
name         0
gender       0
is_smoker    0
party        0
check        0
tip          0
dtype: int64

Answer: TODO

> ### Question 10.  What can you say about the `is_smoker` variable?  I.e., will it bring any insights when analyzing the dataset?  What do you want to do with it?  (and do it...)

In [57]:
# TODO
df.is_smoker.unique()

array([False], dtype=bool)

In [58]:
df.drop('is_smoker', axis=1, inplace=True)


Answer: TODO

> ### Question 11.  For which week days does the dataset has data for?

In [59]:
# TODO
df.day.unique()

array(['Sunday', 'Saturday', 'Thursday', 'Friday'], dtype=object)

Answer: TODO

> ### Question 12.  How often was the bistro patronized for each week day?

(check `.value_counts()`; it could come in handy)

(http://pandas.pydata.org/pandas-docs/stable/generated/pandas.Series.value_counts.html)

In [60]:
# TODO
df.day.value_counts()

Saturday    87
Sunday      76
Thursday    62
Friday      19
Name: day, dtype: int64

Answer: TODO

> ### Question 13.  How much tip did waiters collect for each week day?

In [61]:
# TODO
df.groupby('day').tip.sum()

day
Friday       51.96
Saturday    260.40
Sunday      247.39
Thursday    171.83
Name: tip, dtype: float64

Answer: TODO

> ### Question 14.  What is the average tip per check (in absolute \$) for each week day?

In [62]:
# TODO
df.groupby('day').tip.mean()

day
Friday      2.734737
Saturday    2.993103
Sunday      3.255132
Thursday    2.771452
Name: tip, dtype: float64

Answer: TODO

> ### Question 15.  What is the average tip per check (as a percentage of the check) for each week day?

In [63]:
# TODO
df.groupby('day').tip.sum() / df.groupby('day').check.sum()

day
Friday      0.159445
Saturday    0.146424
Sunday      0.152038
Thursday    0.156732
dtype: float64

Answer: TODO

> ### Question 16.  Are there any name in common between male and female patrons?  (E.g., `Chris` can refer to either a man or a woman)

(check `numpy.intersect1d()`; it could come in handy)

(https://docs.scipy.org/doc/numpy/reference/generated/numpy.intersect1d.html)

In [71]:
# TODO

# Find intersection of two arrays
np.intersect1d([1,2,3],[2,3,4,5])
male_names   = df[df.gender == 'Male'].name # column of male names?
female_names = df[df.gender == 'Female'].name 

np.intersect1d(male_names, female_names) # Values that lie in the intersection

array(['Casey'], dtype=object)

Answer: TODO

> ### Question 17.  If no patrons share the same name, how many unique patrons are in the dataset?

In [75]:
# TODO

# count of men name
men_name_count = len(df[df.gender == 'Male'].name.unique())
# count of women name
women_name_count = len(df[df.gender == 'Female'].name.unique())

men_name_count + women_name_count

182

Answer: TODO

> ### Question 18.  How many times did `Kevin` patronized the bistro?  How about `Alice`?

In [80]:
# TODO
(df.name == 'Kevin').sum()

4

In [81]:
# TODO
(df.name == 'Alice').sum()

2

Answer: TODO

> ### Question 19.  Who are the top 3 female and male patrons?

In [83]:
df[df.gender=='Female'].name.value_counts().head(3) # get top names sorted

Mary     4
Casey    3
Laura    3
Name: name, dtype: int64

In [84]:
df[df.gender =='Male'].name.value_counts().head(3)

David    8
James    5
Casey    5
Name: name, dtype: int64

Answer: TODO

> ### Question 20.  Who's the best tipper (as a fraction of all tips over all check totals)?  Who's the worst?  How many times did they patronize the bistro?

- (http://pandas.pydata.org/pandas-docs/stable/generated/pandas.DataFrame.sort_values.html)
- (http://pandas.pydata.org/pandas-docs/stable/generated/pandas.Series.sort_values.html)

In [91]:
# TODO

# Get all tip amounts
total_tips = df.groupby(['name', 'gender']).tip.sum()
total_checks = df.groupby(['name', 'gender']).check.sum()

(total_tips/total_checks).sort_values(ascending = False)

name      gender
Maryann   Female    0.416667
Bailey    Female    0.325733
Daniel    Male      0.313607
Dennis    Male      0.280535
Zackary   Male      0.266312
                      ...   
Willie    Male      0.066534
Kimberly  Female    0.059447
Destiny   Female    0.056797
Mildred   Female    0.056433
Jeremy    Male      0.035638
Length: 182, dtype: float64

In [92]:
df[df.name == 'Maryann']

day    time     name  gender  party  check  tip
178  Sunday  Dinner  Maryann  Female      2    9.6  4.0

In [ ]:
df[df.name]

Answer: TODO